shared/AppInsightsCore/src/JavaScriptSDK/ChannelController.ts

```diff
@@ -1,135 +1,177 @@
                 }
             });
-        }
-        
-        function _addChannelQueue(queue:IChannelControls[]) {
-            if (queue && queue.length > 0) {
-                queue = queue.sort((a, b) => { // sort based on priority within each queue
-                    return a.priority - b.priority;
+        },
+        isInitialized: () => { return isInitialized },
+        processTelemetry: (item: ITelemetryItem, itemCtx: IProcessTelemetryContext) => {
+            _processChannelQueue(itemCtx, (chainCtx: IProcessTelemetryContext) => {
+                chainCtx.processNext(item);
+            }, () => {
+                itemCtx.processNext(item);
+            });
+        },
+        pause: () => {
+            _processChannelQueue(null, (chainCtx: IProcessTelemetryContext) => {
+                chainCtx.iterate<IChannelControls>((plugin) => {
+                    plugin.pause && plugin.pause();
                 });
-        
-                _checkQueuePriority(queue);
-                _channelQueue.push(queue);
-            }
-        }
-        
-        function _createChannelQueues(channels:IChannelControls[][], extensions: IPlugin[]) {
-            _channelQueue = [];
-        
-            if (channels) {
-                // Add and sort the configuration channel queues
-                arrForEach(channels, queue => _addChannelQueue(queue));
+            });
+        },
+        resume: () => {
+            _processChannelQueue(null, (chainCtx: IProcessTelemetryContext) => {
+                chainCtx.iterate<IChannelControls>((plugin) => {
+                    plugin.resume && plugin.resume();
+                });
+            });
+        },
+        teardown: () => {
+            _processChannelQueue(null, (chainCtx: IProcessTelemetryContext) => {
+                chainCtx.iterate<IChannelControls>((plugin) => {
+                    plugin.teardown && plugin.teardown();
+                });
+            });
+        },
+        flush: (isAsync: boolean, callBack?: () => void) => {
+            let doneIterating = false;
+            let waiting = 0;
+
+            function doCallback() {
+                if (doneIterating && waiting === 0) {
+                    callBack && callBack();
+                }
             }
-        
-            if (extensions) {
-                // Create a new channel queue for any extensions with a priority > the ChannelControllerPriority
-                let extensionQueue:IChannelControls[] = [];
-                arrForEach(extensions as IChannelControls[], plugin => {
-                    if (plugin.priority > ChannelControllerPriority) {
-                        extensionQueue.push(plugin);
+
+            _processChannelQueue(null, (chainCtx: IProcessTelemetryContext) => {
+                chainCtx.iterate<IChannelControls>((plugin) => {
+                    if (plugin.flush) {
+                        waiting ++;
+                        plugin.flush(isAsync, () => {
```

Comment:
Investigation note: What to do if a channel doesn't take callback function or call it!!!

Reply:
Will need to re-work this as the default "Sender" for AI actually doesn't handle any of the arguments for `flush()` so will probably need to add a timeout or just return once all of the channels have had their `flush()` calls triggered...

Reply:
Separating the ChannelController to be a more specific IChannelControls instance so that it can be passed a timeout, also changing the existing `flush` so the functions can return an optional boolean to indicate that they "will" call the callback. This way if the existing flush returns nothing or false the controller will start a timer (defaulting to 5 seconds) to "wait" for all channels to finish flushing before calling any callback. This is SUPER important for a later PR that will correctly handle unloading the entire SDK.